### DISCUSSION TITLE
Slowdown when adding MaterialRankTwoTensorAux auxillary kernel in a peridynamic simulation

### DISCUSSION THREAD

**Discussion by andrisfreimanis** [idaholab/moose#24086] [Q&A Modules: General] [answered]
Dear All,
I was running a simple PD simulation and noticed that adding an auxiliary kernel to output stress slows down the simulation from 38s to 72s. As I understand stress variable is already calculated and MaterialRankTwoTensorAux just accesses the calculated material property, and writes it in the output file. Is something else going on (like some extra calculations)?

The simulation is a simple elastic simulation. 500 nodes and ran on a single core. I attached both input files. They are the same except  that the slower one has  
`  [stress_yy]
    type = MaterialRankTwoTensorAux
    property = stress
    variable = stress_yy
    i = 1
    j = 1
  []`
added to aux kernel block.

[NOSB_elastic.zip](https://github.com/idaholab/moose/files/11247397/NOSB_elastic.zip)

**Comment by hugary1995** [comment, marked as answer]
Well the moose material models are evaluated on-thy-fly, with the very fundamental assumption that the material models are cheap. In other words, every time we compute aux vars, the materials are "re-evaluated" for all the elements.

So if your material models are expensive, for example in a PD simulation, you will want to avoid excessive aux var evaluations as well. Setting `execute_on = TIMESTEP_END` in the auxkernel should make the timing look better. The default execute_on for auxkernels is `LINEAR TIMESTEP_END`.

**Reply by andrisfreimanis** [reply]
Yea, that makes sense. I guess it's the extra evaluation that adds the time.

Thanks
